### ISSUE TITLE
Quantized KV cache (--kv-bits) degenerates into single-token repetition on hybrid MoE models (Qwen3.5/3.6)

### ISSUE THREAD

**Issue by gruenewald** [Blaizzy/mlx-vlm#1322]
First off — thank you for mlx-vlm. It's been fantastic for running VLMs locally on Apple Silicon, and it's the backbone of our document entity-extraction pipeline (German business documents, served through an OpenAI-compatible gateway). We tracked a quality regression from 0.5.0 → 0.6.2 down to a concrete root cause and a reproducible red→green test, and wanted to share it in case it's useful. 🙏

## Context: entity-extraction quality regressed from 0.5.0 → 0.6.2

Our production setup runs `mlx-community/Qwen3.6-35B-A3B-4bit` greedily (`temperature=0`) with **KV-cache quantization enabled** (it both saves memory *and*, on 0.5.0, improved extraction accuracy). On our 6-document / 28-ID extraction suite:

| | 0.5.0 (KV-quant) | 0.6.2 (KV-quant) |
|---|---|---|
| IDs extracted | **28 / 28** | **degenerate output** — single-token repetition (`"9999…"`), effectively 0 usable |

(For completeness: a separate vision-specific recall regression, filed in #1323, accounts for the rest of the gap once this degeneration is out of the way.)

This issue is the cleaner, fully self-contained half of that regression. It reproduces **text-only**, so no model-specific data is needed.

## Summary

With KV-cache quantization (`--kv-bits N`) on the hybrid MoE model `mlx-community/Qwen3.6-35B-A3B-4bit` (`model_type: qwen3_5_moe`), greedy decoding collapses into single-token repetition — e.g. `"9999…"` until `max_tokens`. 0.5.0 produced correct output with the same quantization, so this is a regression.

## Environment (exact)

| | |
|---|---|
| Model | `mlx-community/Qwen3.6-35B-A3B-4bit` (public; `qwen3_5_moe`, 4-bit affine, group_size 64, hidden 2048, 40 layers) |
| mlx-vlm | **0.6.2** / current `main` @ `6d8d5e3` — broken; **0.5.0** — correct |
| mlx / mlx-lm / mlx-metal | 0.31.2 / 0.31.3 / 0.31.2 · Python 3.13 |
| Hardware / OS | Apple M4 Max, macOS 26.5.1 |

Enabling KV quantization differs by version: **0.6.2/main** uses the `--kv-bits` CLI flag; **0.5.0** has no such flag and reads the `KV_BITS` env var. Both end up using a quantized batch KV cache.

## Reproduction — GREEN on 0.5.0, RED on 0.6.2/main (text-only, greedy, deterministic)

```bash
# 0.6.2 / main (RED)
python -m mlx_vlm.server --model mlx-community/Qwen3.6-35B-A3B-4bit --port 8081 --kv-bits 8
# 0.5.0 (GREEN reference — shows it's a regression, not expected behaviour)
KV_BITS=8 python -m mlx_vlm.server --model mlx-community/Qwen3.6-35B-A3B-4bit --port 8081
```
Same greedy request to each:
```json
{"model": "mlx-community/Qwen3.6-35B-A3B-4bit",
 "messages": [{"role": "user", "content": "Repeat the following exactly once: 9988776655"}],
 "temperature": 0.0, "max_tokens": 40}
```

| prompt | 0.5.0 + `KV_BITS=8` (GREEN) | 0.6.2 + `--kv-bits 8` (RED) |
|---|---|---|
| `Repeat the following exactly once: 9988776655` | `9988776655` — `finish_reason=stop` | `9999999999…`×40 — `finish_reason=length` |
| `List the numbers from 1 to 20.` | `1, 2, 3, 4, 5, 6, 7, 8, …` | `1, 2, 3, 4, 5, 7, 7, 7, …` (loops) |

Control: on 0.6.2 **without** `--kv-bits` the same prompts are correct → the degeneration is specific to the quantized KV path.

## Likely root cause

`BatchQuantizedKVCache` (`mlx_vlm/models/cache.py`) doesn't implement `extract()` / `merge()`, which the hybrid per-row decode path relies on:

```
Qwen3_5Model.__call__  ->  _extract_row_cache(cache, row)  ->  cache.extract(row)
                       ...  cache[i] = cache[i].__class__.merge([row_cache])
```

`BatchKVCache` implements both, so the non-quantized path round-trips. For the quantized batch cache, `_extract_row_cache` (`mlx_vlm/models/qwen3_5/language.py`) finds no `extract` method and falls back to an **empty `KVCache`**, dropping the prompt's keys/values and resetting the offset to 0 — so every decode step re-attends from scratch at position 0 → identical logits → the same argmax token each step.

## Red test (synthetic — no model/weights, CI-friendly)

Verified: **2 failed on `main` @ `6d8d5e3`, 2 passed with the fix below.** Drop into `mlx_vlm/tests/test_batch_quantized_cache.py`:

```python
import mlx.core as mx
from mlx_vlm.models.cache import BatchQuantizedKVCache
from mlx_vlm.models.qwen3_5.language import _extract_row_cache

H, D, GROUP, BITS, SEQ = 4, 64, 32, 8, 10

def test_extract_row_cache_preserves_quantized_prompt():
    cache = BatchQuantizedKVCache([0], group_size=GROUP, bits=BITS)
    cache.update_and_fetch(mx.random.normal((1, H, SEQ, D)),
                           mx.random.normal((1, H, SEQ, D)))
    mx.eval(cache.keys)
    row = _extract_row_cache(cache, 0)
    assert int(row.offset) == SEQ              # RED on main: 0 (empty fallback cache)
    assert row.keys is not None and getattr(row, "bits", None) == BITS

def test_extract_merge_roundtrip_quantized():
    cache = BatchQuantizedKVCache([0, 0], group_size=GROUP, bits=BITS)
    cache.update_and_fetch(mx.random.normal((2, H, SEQ, D)),
                           mx.random.normal((2, H, SEQ, D)))
    mx.eval(cache.keys)
    rows = [_extract_row_cache(cache, i) for i in range(2)]
    merged = BatchQuantizedKVCache.merge(rows)   # RED on main: AttributeError (no merge)
    assert merged._idx == SEQ and merged.offset.tolist() == [SEQ, SEQ]
    dq_m = mx.dequantize(*merged.keys, group_size=GROUP, bits=BITS)
    dq_o = mx.dequantize(*cache.keys, group_size=GROUP, bits=BITS)[:, :, :SEQ]
    assert mx.allclose(dq_m, dq_o).item()
```

(Note: `_extract_row_cache` is 0.6.x-only, so this targets the fix rather than a 0.5.0 baseline. The version-level regression is the end-to-end table above.)

## Suggested fix

Implement `BatchQuantizedKVCache.extract()` (mirroring `BatchKVCache.extract`, returning a single-sequence `QuantizedKVCache`) and `merge()` (re-batching single-sequence quantized caches, left-padding to the longest), and have `_extract_row_cache` return an empty `QuantizedKVCache` (not a plain `KVCache`) for an empty quantized batch cache. With that, the synthetic tests pass and the end-to-end repro returns `9988776655`.

I have a working patch + these tests on a branch and would be glad to open a PR if that's helpful — whatever fits your workflow best. Thanks again for the great library!

**Comment by lucasnewman** [comment]
This should have been resolved in https://github.com/Blaizzy/mlx-vlm/pull/1313 -- can you re-test from the main branch? Thanks!

**Comment by lucasnewman** [comment]
I verified your use cases are fixed on main.

**Comment by gruenewald** [comment]
Thanks @lucasnewman — confirmed on my end as well. I re-tested on `main` (`bafd6ce`) with `--kv-bits 8`: the degeneration is gone, greedy `"Repeat the following exactly once: 9988776655"` now returns `9988776655` (`finish_reason=stop`) instead of the `9999…` repetition, and the entity-extraction suite no longer hits the token cap. #1313 resolves it cleanly — nicely done. 🙏

For completeness: the separate, vision-specific recall regression I filed in #1323 is still open on `main` (greedy image extraction recovers fewer IDs than 0.5.0), if you get a chance. But this one's resolved — thanks for the quick turnaround and the great library!
